subt/main.py

```diff
@@ -469,6 +469,32 @@ def on_pose2d(self, timestamp, data):
         self.xyz = x, y, z
         self.trace.update_trace(self.xyz)
 
+    def on_pose3d(self, timestamp, data):
+        xyz, rot = data
+        ypr = quaternion.euler_zyx(rot)
+        pose = (xyz[0], xyz[1], ypr[0])
+        if self.last_position is not None:
+            self.is_moving = (self.last_position != pose)
+            dist = math.hypot(pose[0] - self.last_position[0], pose[1] - self.last_position[1])
+            direction = ((pose[0] - self.last_position[0]) * math.cos(self.last_position[2]) +
+                         (pose[1] - self.last_position[1]) * math.sin(self.last_position[2]))
+            if direction < 0:
+                dist = -dist
+        else:
+            dist = 0.0
+        self.last_position = pose
+        self.traveled_dist += dist
+        x, y, z = xyz
+        self.last_send_time = self.bus.publish('pose2d', [round(x * 1000), round(y * 1000),
+                                    round(math.degrees(self.yaw) * 100)])
+        if self.virtual_bumper is not None:
+            if self.is_virtual:
+                self.virtual_bumper.update_pose(timedelta(seconds=self.sim_time_sec), pose)
+            else:
+                self.virtual_bumper.update_pose(self.time, pose)
+        self.xyz = tuple(xyz)
+        self.trace.update_trace(self.xyz)
```

Comment:
To make sure: When we follow the trace on the way back home, does `self.offset` play any role?

Reply:
as I see it now it does not (and `on_pose3d` would add with offset, while `on_pose2d` without offset) ... but as long only one of them is used all the time, it should be consistent

Reply:
Cool. Thanks for checking.